core/devices/machine_topology.cpp

```diff
@@ -0,0 +1,349 @@
+        if (objects[id].gp_id == gp_index) {
+            return id;
+        }
+    }
+    return -1;
+#endif
+}
+
+
+void MachineTopology::load_objects(
+    hwloc_obj_type_t type, std::vector<MachineTopology::io_obj_info> &vector)
+{
+#if GKO_HAVE_HWLOC
+    GKO_ASSERT(std::get<0>(this->cores_).size() != 0);
+    GKO_ASSERT(std::get<0>(this->pus_).size() != 0);
+    unsigned num_objects = hwloc_get_nbobjs_by_type(this->topo_.get(), type);
+    for (unsigned i = 0; i < num_objects; i++) {
+        // Get the actual PCI object.
+        hwloc_obj_t obj = hwloc_get_obj_by_type(this->topo_.get(), type, i);
+        // Get the non-IO ancestor (which is the closest and the one that can be
+        // bound to) of the object.
+        auto ancestor = hwloc_get_non_io_ancestor_obj(this->topo_.get(), obj);
+        // Create the object.
+        vector.push_back(
+            io_obj_info{obj, obj->logical_index, obj->os_index, obj->gp_index,
+                        hwloc_bitmap_first(ancestor->nodeset), ancestor});
+        // Get the corresponding cpuset of the ancestor nodeset
+        hwloc_cpuset_t ancestor_cpuset = hwloc_bitmap_alloc();
+        hwloc_cpuset_from_nodeset(this->topo_.get(), ancestor_cpuset,
+                                  ancestor->nodeset);
+        // Find the cpu object closest to this device from the ancestor cpuset
+        // and store its id for binding purposes
+        vector.back().closest_cpu_id = get_obj_local_id_by_os_index(
+            std::get<0>(this->pus_), hwloc_bitmap_first(ancestor_cpuset));
+        // Get local id of the ancestor object.
+        if (hwloc_compare_types(ancestor->type, HWLOC_OBJ_PACKAGE) == 0) {
+            vector.back().ancestor_local_id = get_obj_local_id_by_gp_index(
+                std::get<0>(this->packages_), ancestor->gp_index);
+        } else if (hwloc_compare_types(ancestor->type, HWLOC_OBJ_CORE) == 0) {
+            vector.back().ancestor_local_id = get_obj_local_id_by_gp_index(
+                std::get<0>(this->cores_), ancestor->gp_index);
+        } else if (hwloc_compare_types(ancestor->type, HWLOC_OBJ_NUMANODE) ==
+                   0) {
+            vector.back().ancestor_local_id = get_obj_local_id_by_gp_index(
+                std::get<0>(this->numa_nodes_), ancestor->gp_index);
+        }
+        hwloc_bitmap_free(ancestor_cpuset);
+        // Get type of the ancestor object and store it as a string.
+        char ances_type[24];
+        hwloc_obj_type_snprintf(ances_type, sizeof(ances_type), ancestor, 0);
+        vector.back().ancestor_type = std::string(ances_type);
+        // Write the PCI Bus ID from the object info.
+        char pci_bus_id[13];
+        snprintf(pci_bus_id, sizeof(pci_bus_id), "%04x:%02x:%02x.%01x",
+                 obj->attr->pcidev.domain, obj->attr->pcidev.bus,
+                 obj->attr->pcidev.dev, obj->attr->pcidev.func);
+        vector.back().pci_bus_id = std::string(pci_bus_id);
```

Comment:
You are adding a lot of information to this object. Aren't most of them unneeded, except for this portion? What comes after this code definitely looks like we do not need it for now.

I would rather we make this function more slim for now and only use what we need, and later if we need extra information for whichever reason we can always add them.

Reply:
Initially, I planned to use the names of the io children objects to get the type of devices. I think that still makes sense for detecting network cards or Infiniband cards, as I am not sure how to detect those with just a PCI id. But I guess as currently we dont have any network based support, we can add it later when necessary.